feat: earliest available slot

matthewkeil opened the pull request:
**Motivation**

Implement `earliest_available_slot` from ethereum/consensus-specs/pull/4374

matthewkeil committed:
feat: add lowestSlot event and emit from chain at init

matthewkeil committed:
feat: add fulu.Status to types package

matthewkeil committed:
feat: add fulu.Status ssz types to reqresp config

matthewkeil committed:
feat: add StatusV2 to reqresp

matthewkeil committed:
feat: add notification to update status from chain

matthewkeil committed:
chore: fix-types for ReqRespBeaconNode

matthewkeil committed:
feat: update network to pass fulu.Status onUpdateStatus

matthewkeil committed:
feat: update statusCache and network consumers to use new Status

matthewkeil committed:
chore: lint

matthewkeil committed:
chore: remove unused comment

matthewkeil committed:
feat: update all peer related types to multifork Status

matthewkeil committed:
docs: add comments and line space

matthewkeil committed:
feat: add and implement PeerSyncMeta

matthewkeil committed:
feat: implement PeerSyncMeta and getConnectedPeerSyncMeta

nflaig commented on packages/beacon-node/src/network/network.ts:
```diff
@@ -699,17 +690,25 @@ export class Network implements INetwork {
-    this.connectedPeers.set(data.peer, data.dataColumns);
-    this.connectedPeerClients.set(data.peer, data.clientAgent);
+    const earliestAvailableSlot = (data.status as fulu.Status).earliestAvailableSlot;
+    this.logger.warn("onPeerConnected", {
```
why is this a `warn`, just for early debugging?

matthewkeil replied:
that is what it was.  changed to `verbose`

nflaig commented on packages/beacon-node/src/network/reqresp/types.ts:
```diff
@@ -99,7 +110,7 @@ const blocksResponseType: ResponseTypeGetter<SignedBeaconBlock> = (fork, version
 
 export const responseSszTypeByMethod: {[K in ReqRespMethod]: ResponseTypeGetter<ResponseBodyByMethod[K]>} = {
-  [ReqRespMethod.Status]: () => ssz.phase0.Status,
+  [ReqRespMethod.Status]: (_, version) => (version === Version.V2 ? ssz.fulu.Status : ssz.phase0.Status),
```
why do we use the version instead of `fork` (first param `_`) to determine type?

**Edit:** noticed it's same for `Metadata` is it because we use them before fork already? might wanna leave a comment here if that's the case

matthewkeil replied:
yes. correct, its based on sender request not on the fork

nflaig reviewed (commented):
Looks good in general, opened a PR with some simplifications https://github.com/ChainSafe/lodestar/pull/7987

nflaig committed:
chore: review #7959 (#7987)

Mostly simplified how we pick `Status` type by using `sszTypesFor`, this
also makes it more forward compatible in case we change the type in the
future again.

matthewkeil committed:
fix: change log to verbose

twoeths commented on packages/beacon-node/src/chain/chain.ts:
```diff
@@ -188,6 +188,16 @@ export class BeaconChain implements IBeaconChain {
+
+  set earliestSlotAvailable(slot: Slot) {
+    this._earliestSlotAvailable = slot;
+    this.emitter.emit(ChainEvent.updateStatus);
```
maybe check with the old value before `emit`?

twoeths commented on packages/beacon-node/src/network/peers/utils/assertPeerRelevance.ts:
```diff
@@ -19,11 +19,7 @@ type IrrelevantPeerType =
-  local: phase0.Status,
-  currentSlot: Slot
-): IrrelevantPeerType | null {
+export function assertPeerRelevance(remote: Status, local: Status, currentSlot: Slot): IrrelevantPeerType | null {
```
post fulu I think it's required to have `cgc` + `earliest_available_slot`
we either need to disconnect these peers or accept it and come up with tracking how many peers missing these values in metrics
this will help us investigate any peering issues later

matthewkeil replied:
did check for status but need to add for metadata.cgc under separate PR. created an issue here
https://github.com/ChainSafe/lodestar/issues/7999

twoeths commented on packages/beacon-node/src/sync/range/utils/peerBalancer.ts:
```diff
@@ -29,20 +28,26 @@ export class ChainPeersBalancer {
     const sortedBestPeers = sortBy(
-      this.peers.filter((peerId) => {
+      this.peers.filter(({earliestAvailableSlot, custodyGroups}) => {
+        const earliestSlot = earliestAvailableSlot ?? 0;
```
this means we optimistically pick a peer without earliest_available_slot
if that's the case we need to deprioritize these peers, ie if we have peers with `earliest_available_slot` with other conditions we should pick it

that could be a TODO for #7746, we refactor the logic in that PR anyway

matthewkeil replied:
ooooo!!! really good catch!!  I added a TODO to the code at that location.

matthewkeil replied:
https://github.com/ChainSafe/lodestar/pull/7746#issuecomment-2998148642

twoeths commented on packages/beacon-node/src/chain/chain.ts:
```diff
@@ -294,6 +304,7 @@ export class BeaconChain implements IBeaconChain {
             pubkey2index: new PubkeyIndexMap(),
             index2pubkey: [],
           });
+    this._earliestSlotAvailable = cachedState.slot;
```
add a TODO so that when we implement backfill sync, we update this value gradually

matthewkeil replied:
https://github.com/ChainSafe/lodestar/issues/7997

twoeths commented on packages/beacon-node/src/network/reqresp/ReqRespBeaconNode.ts:
```diff
@@ -224,7 +229,7 @@ export class ReqRespBeaconNode extends ReqResp {
     const protocolsAtFork: [ProtocolNoHandler, ProtocolHandler][] = [
       [protocols.Ping(fork, this.config), this.onPing.bind(this)],
-      [protocols.Status(fork, this.config), this.onStatus.bind(this)],
+      [protocols.StatusV2(fork, this.config), this.onStatus.bind(this)],
```
may want to move this to after `MetadataV3` to make it easier to maintain why we add StatusV2 by default
or add some more comments for it similar to `MetadataV3`

twoeths commented on packages/beacon-node/src/network/reqresp/ReqRespBeaconNode.ts:
```diff
@@ -347,6 +357,8 @@ export class ReqRespBeaconNode extends ReqResp {
 
     const metadata = this.metadataController.json;
+
+    // TODO(fulu): is this still correct?  there are three versions of Metadata now...
```
I think it's clearly a bug, need to open an issue for it

matthewkeil replied:
https://github.com/ChainSafe/lodestar/issues/7998

twoeths commented on packages/beacon-node/src/network/reqresp/beaconBlocksMaybeBlobsByRange.ts:
```diff
@@ -78,7 +79,12 @@ export async function beaconBlocksMaybeBlobsByRange(
     const neededColumns = partialDownload ? partialDownload.pendingDataColumns : sampledColumns;
-    const peerColumns = network.getConnectedPeerCustody(peerId);
+
+    if ((earliestAvailableSlot ?? 0) > startSlot) {
```
add comment noting we already check this in `ChainPeersBalancer` before this step

matthewkeil committed:
fix: check for changed earliest slot value before update and emit

matthewkeil committed:
docs: move StatusV2 protocol at fork init next to metadata

matthewkeil committed:
docs: add note about peer balancer check

matthewkeil committed:
feat: down score peer if not announced earliest available slot

matthewkeil committed:
docs: add comment about bug in peerBalancer that prioritizes peers incorrectly

matthewkeil committed:
Merge branch 'peerDAS' into mkeil/earliest-available-slot

matthewkeil commented:
@ensi321  can you double check this small merge conflict I resolved.

https://github.com/ChainSafe/lodestar/pull/7959/commits/645363cc30e46d7418c4292988505a63277a564e

The only question was in the onStatus handler in 

packages/beacon-node/src/network/reqresp/ReqRespBeaconNode.ts

on line 334

ensi321 commented:
> @ensi321 can you double check this small merge conflict I resolved.
> 
> [645363c](https://github.com/ChainSafe/lodestar/commit/645363cc30e46d7418c4292988505a63277a564e)
> 
> The only question was in the onStatus handler in
> 
> packages/beacon-node/src/network/reqresp/ReqRespBeaconNode.ts
> 
> on line 334

Looks good

wemeetagain commented:
:tada: This PR is included in [v1.34.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.34.0) :tada:
